File: DigitalImageProcessing/python/sobel_operator.py
#!/usr/bin/env python3
# encoding: utf-8

"""
@Funciton: Sobel 算子进行边缘检测 —— 可分离卷积核
@Python Version: 3.8
@Author: Wei Li
@Date: 2021-10-14
"""

import sys
import math
from scipy import signal
import numpy as np
import cv2 as cv


def PascalSmooth(n):
    """函数 PascalSmooth 返回 n 阶的非归一化的高斯平滑算子，
    即指数为 n-1 的二项式展开式的系数，
    其中对于阶乘的实现，利用 Python 的函数包 math 中的 factorial，其参数 n 为奇数。

    Args:
        n ([int]): 高斯卷积算子的阶数(奇数)

    Returns:
        [array]: 高斯卷积算子中的系数，即用于 Soble 算子中平滑核参数
    """
    pascalSmooth = np.zeros([1, n], np.float32)
    for idx in range(n):
        pascalSmooth[0][idx] = math.factorial(n - 1) / math.factorial(idx) * math.factorial(n - 1 - idx)

    return pascalSmooth


def PascalDiff(n):
    """函数 PascalDiff 返回 n 阶差分算子，完成 Sobel 在方向上的差分操作

    Args:
        n ([int]): Sobel 进行 n 阶差分

    Returns:
        [array]: Soble n 阶差分结果
    """
    pascalDiff = np.zeros([1, n], np.float32)
    pascalSmooth_previous = PascalSmooth(n - 1)
    for idx in range(n):
        if idx == 0:
            # 恒等于 1
            pascalDiff[0][idx] = pascalSmooth_previous[0][idx]
        elif idx == n - 1:
            # 恒等于 -1
            pascalDiff[0][idx] = -pascalSmooth_previous[0][idx - 1]
        else:
            pascalDiff[0][idx] = pascalSmooth_previous[0][idx] - pascalSmooth_previous[0][idx - 1]

    return pascalDiff


def GetSobelKernel(n):
    """ PascalSmooth 返回的平滑算子和 PascalDiff 返回的差分算子进行 full 卷积，
    就可以得到完整的水平方向和垂直方向上的 nxn 的 Sobel 算子。
    注意，真正在进行 Sobel 卷积时，这一步是多余的，直接通过卷积的分离性就可以完成 Sobel 卷积,
    这里只是为了得到完整的 Sobel 核,通过定义函数 getSobelKernel 来实现, 返回值包括水平方向和垂直方向上的 Sobel 核。

    Args:
        n ([int]): n 阶 Sobel 算子

    Returns:
        [array]: 水平方向和垂直方向的 Sobel 卷积核
    """
    pascalSmoothKernel = PascalSmooth(n)
    pascalDiffKernel = PascalDiff(n)

    # 水平方向上卷积核
    sobelKernel_x = signal.convolve2d(pascalSmoothKernel.transpose(), pascalDiffKernel, mode="full")

    # 垂直方向上卷积核
    sobelKernel_y = signal.convolve2d(pascalSmoothKernel, pascalDiffKernel.transpose(), mode="full")

    return (sobelKernel_x, sobelKernel_y)


def SobelOperator(image, n):
    """ 构建了 Sobel 平滑算子和差分算子后，通过这两个算子来完成图像矩阵与 Sobel 算子的 same 卷积，
    函数 SobelOperator 实现该功能: 
        图像矩阵先与垂直方向上的平滑算子卷积得到的卷积结果，
        再与水平方向上的差分算子卷积，
        这样就得到了图像矩阵与sobel_x 核的卷积。
        与该过程类似,图像矩阵先与水平方向上的平滑算子卷积得到的卷积结果,
        再与垂直方向上的差分算子卷积,
        这样就得到了图像矩阵与 sobel_y 核的卷积。

    Args:
        image ([ndarray]): 进行 Sobel 算子的原始输入图像
        n ([int]): 进行 Sobel 算子的阶数

    Returns:
        [ndarray]: 水平方向上的 Sobel 卷积结果；垂直方向上的卷积结果
    """
    pascalSmoothKernel = PascalSmooth(n)
    pascalDiffKernel = PascalDiff(n)

    # -------- 与水平方向上 Sobel 卷积核进行卷积 --------
    # 可分离卷积核 1. 先进行垂直方向的平滑
    img_sobel_x = signal.convolve2d(image, pascalSmoothKernel.transpose(), mode="same")
    # 可分离卷积核 2. 再进行水平方向的差分
    img_sobel_x = signal.convolve2d(img_sobel_x, pascalDiffKernel, mode="same")

    # -------- 与水平方向上 Sobel 卷积核进行卷积 --------
    # 可分离卷积核 1. 先进行垂直方向的平滑
    img_sobel_y = signal.convolve2d(image, pascalSmoothKernel, mode="same")
    # 可分离卷积核 2. 再进行水平方向的差分
    img_sobel_y = signal.convolve2d(img_sobel_x, pascalDiffKernel.transpose(), mode="same")

    return img_sobel_x, img_sobel_y


# --------------------------
if __name__ == "__main__":
    if len(sys.argv) > 1:
        image = cv.imread(sys.argv[1], 0)
        if image is None:
            print(f"Error: no such file or dictory.")
        cv.imshow("OriginImage",image)

        # 注意区分边缘强度和边缘强度的灰度级显示
        img_soble_x, img_soble_y = SobelOperator(image, 3)

        # 计算绝对值，获取水平方向和垂直方向的边缘强度
        abs_img_soble_x = np.abs(img_soble_x)
        abs_img_soble_y = np.abs(img_soble_y)

        # 水平方向和垂直方向的边缘强度的灰度级显示
        edge_x = abs_img_soble_x.copy()
        edge_y = abs_img_soble_y.copy()
        # 将大于 255 直接进行饱和操作
        edge_x[edge_x>255] = 255
        edge_y[edge_y>255] = 255
        # 数据类型转换
        edge_x = edge_x.astype(np.uint8)
        edge_y = edge_y.astype(np.uint8)
        cv.imshow("edge_x", edge_x)
        cv.imshow("edge_y", edge_y)

        # 根据 sobel 两个卷积结果, 计算最终的边缘强度
        # 计算最终的边缘强度有多种方式, 采用 平方和开方方式
        edge = np.sqrt(np.power(img_soble_x, 2.0) + np.power(img_soble_y, 2.0))

        # 边缘轻度的灰度级显示
        # Sobel 边缘检测，将边缘强度大于 255 的值直接截断为 255，这样得到的边缘有可能不够平滑
        # edge[edge>255] = 255
        # edge = edge.astype(np.uint8)
        # cv.imshow("edge_255", edge)

        # 另一种方式，对所得到的边缘强度进行直方图正规化处理或者归一化处理。
        # 对边缘强度进行归一化处理得到边缘强度的灰度级显示，如果得到的对比度较低，还可以通过伽马变换进行对比度增强
        edge = edge / np.max(edge)
        edge = np.power(edge, 1)
        edge *= 255
        edge = edge.astype(np.uint8)
        cv.imshow("Soble_scale", edge)
        # cv.imwrite("./image/Soble_scale.png", edge)

        cv.waitKey(0)
        cv.destroyAllWindows()

    else:
        print("Usge: python.py imageFile")

File: DigitalImageProcessing/cpp/sobel_operator.cpp
/**
 * @File    : sobel_operator.cpp
 * @Brief   : Sobel 算子进行边缘检测  —— 可分离卷积核
 * @Author  : Wei Li
 * @Date    : 2021-10-14
*/

#include <iostream>
#include <opencv2/core.hpp>
#include <opencv2/imgproc.hpp>
#include <opencv2/highgui.hpp>

// 可以将这一系列的函数操作封装为一个类
// 卷积运算 OpenCV with CPlusPlus
void conv2D(cv::InputArray src, cv::InputArray kernel, cv::OutputArray dst, int ddepth, cv::Point anchor=cv::Point(-1, -1), int borderType=cv::BORDER_DEFAULT)
{
    // 1. 卷积核逆时针翻转 180 度
    cv::Mat kernelFlip;
    cv::flip(kernel, kernelFlip, -1);
    // 利用计算自相关函数完成卷积运算
    cv::filter2D(src, dst, ddepth, kernelFlip, anchor, 0.0, borderType);
}

// 可分离卷积运算
// 可分离离散二维卷积运算，先进行垂直方向卷积，再进行水平方向卷积
void sepConv2D_Y_X(cv::InputArray src, cv::OutputArray src_kerY_kerX, int ddepth, cv::InputArray kernelY, cv::InputArray kernelX, cv::Point anchor=cv::Point(-1, -1), int borderType=cv::BORDER_DEFAULT)
{
    cv::Mat src_kerY;
    conv2D(src, kernelY, src_kerY, ddepth, anchor, borderType);
    conv2D(src_kerY, kernelX, src_kerY_kerX, ddepth, anchor, borderType);
}

// 可分离离散二维卷积运算，先进行水平方向卷积，再进行垂直方向卷积
void sepConv2D_X_Y(cv::InputArray src, cv::OutputArray src_kerX_kerY, int ddepth, cv::InputArray kernelX, cv::InputArray kernelY, cv::Point anchor=cv::Point(-1, -1), int borderType=cv::BORDER_DEFAULT)
{
    cv::Mat src_kerX;
    conv2D(src, kernelX, src_kerX, ddepth, anchor, borderType);
    conv2D(src_kerX, kernelY, src_kerX_kerY, ddepth, anchor, borderType);
}


// Factorial 函数实现阶乘
int Factorial(int n)
{
    int fac = 1;
    if (n == 0)
    {
        return fac;
    }
    for (int i = 1; i <= n; ++i)
    {
        fac *= i;
    }

    return fac;
}

// getPascalSmooth 函数实现获取 soble 平滑算子
cv::Mat getPascalSmooth(int n)
{
    cv::Mat pascalSmooth = cv::Mat::zeros(cv::Size(n, 1), CV_32FC1);
    for (int i = 0; i < n; ++i)
    {
        pascalSmooth.at<float>(0, i) = Factorial(n - 1) / (Factorial(i) * Factorial(n - 1 - i));
    }

    return pascalSmooth;
}

// getPascalDiff 函数实现获取 soble 差分算子
cv::Mat getPascalDiff(int n)
{
    cv::Mat pascalDiff = cv::Mat::zeros(cv::Size(n, 1), CV_32FC1);
    cv::Mat pascalSmooth_previous = getPascalSmooth(n - 1);
    for (int i = 0; i < n; ++i)
    {
        if (i == 0) // 恒等于 1
        {
            pascalDiff.at<float>(0, i) = 1;
        }
        else if (i == n - 1) // 恒等于 -1
        {
            pascalDiff.at<float>(0, i) = -1;
        }
        else
        {
            pascalDiff.at<float>(0, i) = pascalSmooth_previous.at<float>(0, i) - pascalSmooth_previous.at<float>(0, i - 1);
        }
    }

    return pascalDiff;
}

// SobelOperator 函数实现图像核 soble 算子的卷积操作
// 当参数 x_flag !=0 时，返回图像矩阵与 sobel_x 核的卷积 ;
// 当 x_flag = 0且 y_flag !=0 时，返回图像矩阵与 sobel_y 核的卷积。
cv::Mat SobelOperator(cv::Mat image, int x_flag, int y_flag, int winSize, int borderType)
{
    // sobel 卷积核的大小为大于 3 的奇数
    CV_Assert(winSize >= 3 && winSize % 2 == 1);
    // 平滑系数
    cv::Mat pascalSmooth = getPascalSmooth(winSize);
    // 差分系数
    cv::Mat pascalDiff = getPascalDiff(winSize);
    cv::Mat img_sobel_conv;

    // ---- x_flag != 0, image and soble_x kernel convolution ----Horizontal
    if (x_flag != 0)
    {
        // 可分离卷积性质：先进行一维垂直方向的平滑，再进行一维水平方向的差分
        sepConv2D_Y_X(image, img_sobel_conv, CV_32FC1, pascalSmooth.t(), pascalDiff, cv::Point(-1, -1), borderType);
    }

    // ---- x_flag != 0, image and soble_x kernel convolution ----Vertical
    if (x_flag == 0 && y_flag != 0)
    {
        // 可分离卷积性质：先进行一维水平方向的平滑，再进行一维垂直方向的差分
        sepConv2D_X_Y(image, img_sobel_conv, CV_32FC1, pascalSmooth, pascalDiff.t(), cv::Point(-1, -1), borderType);
    }

    return img_sobel_conv;
}


// -------------------------------
int main(int argc, char **argv)
{
    cv::Mat image = cv::imread(argv[1], 0);
    if (!image.data)
    {
        std::cout << "Error: Reading image file unsuccessfully.";
        return -1;
    }
    cv::imshow("OriginImage", image);

    // 原始图像和 soble kernerl-x 卷积
    cv::Mat img_sobel_x = SobelOperator(image, 1, 0, 3, cv::BORDER_DEFAULT);

    // 原始图像和 soble kernerl-y 卷积
    cv::Mat img_sobel_y = SobelOperator(image, 0, 1, 3, cv::BORDER_DEFAULT);

    // 两个卷积结果的灰度级显示
    cv::Mat abs_img_sobel_x, abs_img_sobel_y;
    cv::convertScaleAbs(img_sobel_x, abs_img_sobel_x, 1, 0);
    cv::convertScaleAbs(img_sobel_y, abs_img_sobel_y, 1, 0);
    cv::imshow("horizontal_edge", abs_img_sobel_x);
    cv::imshow("vertical_edge", abs_img_sobel_y);

    // 根据 soble 两个卷积结果，获取最终图像的边缘强度
    cv::Mat edge;
    cv::Mat img_sobel_x_x, img_sobel_y_y;
    cv::pow(img_sobel_x, 2.0, img_sobel_x_x);
    cv::pow(img_sobel_y, 2.0, img_sobel_y_y);
    cv::sqrt(img_sobel_x_x + img_sobel_y_y, edge);
    // 数据类型转换，边缘强度的灰度级显示
    edge.convertTo(edge, CV_8UC1);
    cv::imshow("edge", edge);

    // OpenCV 函数，直接计算 sobel 算子卷积结果
    // void Sobel(src, dst, ddepth, dx, dy, ksize, scale, delta, borderType)

    cv::waitKey(0);
    cv::destroyAllWindows();
    return 0;
}
